### ISSUE TITLE
use -fsanitize=undefined when compiling C code in Debug and ReleaseSafe build modes

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#3569]
The diff for this is simple:

```diff
--- a/src/codegen.cpp
+++ b/src/codegen.cpp
@@ -8804,6 +8804,8 @@ void add_cc_args(CodeGen *g, ZigList<const char *> &args, const char *out_dep_pa
 
     switch (g->build_mode) {
         case BuildModeDebug:
+            args.append("-fsanitize=undefined");
+
             // windows c runtime requires -D_DEBUG if using debug libraries
             args.append("-D_DEBUG");
 
@@ -8816,6 +8818,8 @@ void add_cc_args(CodeGen *g, ZigList<const char *> &args, const char *out_dep_pa
             }
             break;
         case BuildModeSafeRelease:
+            args.append("-fsanitize=undefined");
+
             // See the comment in the BuildModeFastRelease case for why we pass -O2 rather
             // than -O3 here.
             args.append("-O2");

```

However this causes clang to emit calls to handler functions, for example:

```
[nix-shell:~/Downloads/zig/build]$ ./zig build-exe --c-source test.c -lc 
Build Dependencies...crtn [2] Compile C Objects [1/1] crtn.S...lld: error: undefined symbol: __ubsan_handle_pointer_overflow
>>> referenced by elf-init.c:88 (/home/andy/Downloads/zig/lib/libc/glibc/csu/elf-init.c:88)
>>>               /home/andy/.local/share/zig/stage1/o/MAQLigNffSWFYhMSJ8sFs_CRyB2W-VPbrPlqEr0pGVmIr_kEWJEqijRGwX7Gn_ga/elf-init.o:(__libc_csu_init) in archive /home/andy/.local/share/zig/stage1/o/bDCPUp34y2-OELrSYdQbDQgP8-KwNaHhfc4JYG5UPI6KsbqojxucV5gs2E-vjPYN/libc_nonshared.a
```

One can see the list of handler functions in LLVM's compiler-rt project:

```
andy@ark ~/D/l/compiler-rt> grep -RI __ubsan_handle
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_add_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_add_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_alignment_assumption)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_alignment_assumption_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_builtin_unreachable)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_cfi_bad_type)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_cfi_check_fail)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_cfi_check_fail_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_divrem_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_divrem_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_dynamic_type_cache_miss)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_dynamic_type_cache_miss_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_float_cast_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_float_cast_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_function_type_mismatch_v1)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_function_type_mismatch_v1_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_implicit_conversion)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_implicit_conversion_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_invalid_builtin)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_invalid_builtin_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_load_invalid_value)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_load_invalid_value_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_missing_return)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_mul_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_mul_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_negate_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_negate_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nonnull_arg)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nonnull_arg_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nonnull_return_v1)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nonnull_return_v1_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nullability_arg)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nullability_arg_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nullability_return_v1)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_nullability_return_v1_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_out_of_bounds)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_out_of_bounds_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_pointer_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_pointer_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_shift_out_of_bounds)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_shift_out_of_bounds_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_sub_overflow)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_sub_overflow_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_type_mismatch_v1)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_type_mismatch_v1_abort)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_vla_bound_not_positive)
lib/ubsan/ubsan_interface.inc:INTERFACE_FUNCTION(__ubsan_handle_vla_bound_not_positive_abort)
```

We can provide our own handlers for these functions, or use `-fsanitize-trap` instead, which requires no runtime library. However this did not seem to catch UB:

```c
#include <stdio.h>

int main(int argc, char **argv) {
    int k = 0x7fffffff;
    k += argc;

    return 0;
}
```

```
$ ./zig build-exe --c-source test.c -lc --verbose-cc
/home/andy/Downloads/zig/build/zig cc -MD -MV -MF zig-cache/tmp/fd41bxBpB4Tb-test.o.d -nostdinc -fno-spell-checking -isystem /home/andy/Downloads/zig/lib/include -isystem /home/andy/Downloads/zig/lib/libc/include/x86_64-linux-gnu -isystem /home/andy/Downloads/zig/lib/libc/include/generic-glibc -isystem /home/andy/Downloads/zig/lib/libc/include/x86_64-linux-any -isystem /home/andy/Downloads/zig/lib/libc/include/any-linux-any -march=native -g -fno-omit-frame-pointer -fsanitize-trap=undefined -D_DEBUG -fstack-protector-strong --param ssp-buffer-size=4 -fPIC -o zig-cache/tmp/fd41bxBpB4Tb-test.o -c test.c
$ ./test
$
```

**Comment by vi** [comment]
Is `-fsanitize` available on all LLVM-supported platforms?

**Comment by andrewrk** [comment]
> Is `-fsanitize` available on all LLVM-supported platforms?

The version that emits a trap is available on all targets. If this turns out not to be true, then it's easy to address this in Zig:

https://github.com/ziglang/zig/blob/13cdc137e6389af83f225ab851bf8ef768ebcc69/src/target.cpp#L1609-L1611

**Comment by noloader** [comment]
@vi,

> Is -fsanitize available on all LLVM-supported platforms?

Undefined Behavior sanitizer (`-fsanitize=undefined`) is available for Clang 3.1 and above. It was available earlier but you had to compile Clang from sources using John Regehr's patches.

The other sanitizers were added over time, like Asan and Msan. I don'trecall when they became available.
